add resume hooks

melvinw opened the pull request:
It would be nice if module writers had a way to perform tasks at resume-time without having to modify the `ResumeAll()` and `ResumeWorker()` RPCs. This set of commits introduces `ResumeHooks` to provide that functionality.

melvinw commented:
Not entirely sure if the interface I came up with this is the cleanest-- definitely willing to rework it if anyone has an opinion on the matter. The `ResumeHookUser` stuff is there mostly to avoid having to scan the entire map of modules on resume, as it can grow to be quite large.

chris3torek commented on core/resume_hook.cc:
```diff
@@ -0,0 +1,141 @@
+  static std::map<std::string, ResumeHookFactory<Worker>>
+      all_resume_hook_factories;
+
+  if (reset) {
```
I can't figure out where reset is ever true (but this is the same as the gate hook code, where I also can't figure out where this is ever true, so at least I'm consistently missing something 😺 )

melvinw replied:
`reset` is here to facilitate testing. `ModuleBuilder::all_module_builders_holder` has a similar flag that I could have sworn used to be used in some tests, but I can't find them anymore either

chris3torek commented on core/gate.cc:
```diff
@@ -72,7 +72,7 @@ int Gate::AddHook(GateHook *hook) {
 
   hooks_.push_back(hook);
-  std::sort(hooks_.begin(), hooks_.end(), GateHookComp);
+  std::sort(hooks_.begin(), hooks_.end());
```
(see later comment about sort-right-after-insert ... I wound up going through this backwards)

melvinw replied:
Good point. Switched to `std::set` over `std::vector` here since this isn't on the fast path.

chris3torek commented on core/bessctl.cc:
```diff
@@ -1376,6 +1463,107 @@ class BESSControlImpl final : public BESSControl::Service {
+
+    uint64_t worker_filter = request->workers();
+    while (worker_filter) {
+      Worker* w = workers[wid];
```
The documentation says that the filter is a bitmap, but we don't actually test the specific bit (`worker_filter & 1`) here...

melvinw replied:
Doh. Thanks for catching-- will fix

chris3torek replied:
Whoops, the `continue` skips the `worker_filter >>= 1` and `wid++` lines 😮

melvinw replied:
Oh, yikes! Fixed and rebased

chris3torek commented on core/bessctl.cc:
```diff
@@ -1241,10 +1327,10 @@ class BESSControlImpl final : public BESSControl::Service {
+        WorkerPauser wp;  // Only pause when absolutely required
+        ret = m1->ConnectModules(ogate, m2, igate);
+        goto done;
       }
```
(general comment: I like having the whitespace fixes as a separate commit - maybe we should just get that one in and done with first)

melvinw replied:
Ah, whichever commit touched these lines last must not have been clang-formated correctly. We have a pre-commit hook for this, but it only really works if everyone installs it :)

chris3torek commented on core/bessctl.cc:
```diff
@@ -531,17 +562,72 @@ class BESSControlImpl final : public BESSControl::Service {
+    ResumeHookUser<Worker>* global_hooks =
+        bess::ResumeHookUser<Worker>::get_user_hooks(nullptr);
+
+    // Enable metadata offset computation by default
```
I wish there were a good way to get this to happen at constructor time, when we build the metadata resume hook... not sure if there is though, we might get into constructor ordering issues.

melvinw replied:
Yeah, I thought about putting this logic in the [Pipeline constructor](https://github.com/NetSys/bess/blob/master/core/metadata.h#L148), but it's not clear how to handle re-adding the hook after a `daemon reset`.

melvinw commented on core/resume_hooks/README.md:
```diff
@@ -0,0 +1,17 @@
+- Define `YourHook::kName`. This must be unique across all other resume hooks.
+- Define `YourHook::kPriority`. Ties are broken by hook name in increasing lexographical order.
+- Define `void YourHook::Run()`.
+
```
I may be able to move this to the `ADD_MODULE_RESUME_HOOK`/`ADD_WORKER_RESUME_HOOK` macros. Will update the docs either way. Thanks

melvinw replied:
Fixed along with some other minor things in this doc

melvinw commented:
Marking this not-ready while I refactor things

sangjinhan commented:
Did some offline discussion for higher bandwidth. One reminder for later: resume hooks should be called also when pause/resume is done implicitly (by WorkerPauser)

melvinw commented:
Rebased and updated to reflect our conversation, @sangjinhan . I also added a possible implementation of a thread-safe `PortOut` as an example of how to implement pre-resume hooks for modules.

sangjinhan commented on core/resume_hook.cc:
```diff
@@ -0,0 +1,70 @@
+namespace bess {
+
+std::set<std::unique_ptr<ResumeHook>> global_resume_hooks;
+std::set<Module *> event_modules;
```
Perhaps should be in `event.cc`, since Resume is just one particular type of events?

sangjinhan commented on core/bessctl.cc:
```diff
@@ -1147,6 +1156,10 @@ class BESSControlImpl final : public BESSControl::Service {
       response->set_name(module->name());
     }
 
+    if (module->OnEvent(bess::Event::PreResume) != ENOTSUP) {
```
This event is somewhat spurious, since `PreResume` event is not really happening here. I think it should be fine for most module handlers, but how about,
1. By default, we add all new modules to `event_modules` (without `if` here).
1. `ENOTSUP` is checked in the loop over `event_modules`.

melvinw replied:
Good point. Will change

sangjinhan commented on core/resume_hook.h:
```diff
@@ -0,0 +1,136 @@
+
+  virtual void Run() = 0;
+
+  bool operator<(const ResumeHook &rhs) {
```
This method can be `const`

sangjinhan commented on core/worker.cc:
```diff
@@ -403,20 +406,59 @@ bool detach_tc(bess::TrafficClass *c) {
   for (int wid : workers_paused_) {
+    for (Module *m : bess::event_modules) {
+      if (!modules_run.count(m) && m->active_workers()[wid]) {
+        m->OnEvent(bess::Event::PreResume);
```
For each module, the event seems to be triggered twice (here and `run_global_resume_hooks` above)

melvinw replied:
Ah, good catch. Will fix

sangjinhan commented on core/worker.cc:
```diff
@@ -403,20 +406,59 @@ bool detach_tc(bess::TrafficClass *c) {
   return remove_tc_from_orphan(c);
 }
 
+void run_global_resume_hooks() {
```
`resume_hook.cc` might be a better place for this function (not sure)

melvinw replied:
Agreed. I'll need to rework this slightly. Moving as is introduces an awkard circular dependency between `resume_hook.h` and `resume_hooks/setup_metadata.h`

sangjinhan commented on protobuf/bess_msg.proto:
```diff
@@ -499,6 +499,17 @@ message ConfigureGateHookRequest {
+// Enable/Disable a global ResumeHook.
+message ConfigureResumeHookRequest {
+  string hook_name = 1;         /// Name of the hook
+  bool enable = 5;              /// Installs the hook if True, else uninstalls
```
Since the protobuf has not been released, we can do 1, 2, 3, ... for the member index?

melvinw replied:
Definitely

sangjinhan commented on pybess/bess.py:
```diff
@@ -495,6 +495,15 @@ def _configure_gate_hook(self, hook, module,
         request.arg.Pack(arg)
         return self._request('ConfigureGateHook', request)
 
+    def _configure_resume_hook(self, hook, arg, enable=None):
```
Since the method is meant to be used externally, the method name should not begin with `_`.

(`enable=True` should be fine...?)

melvinw replied:
Ah, yup-- but I think we still need the `is None` check below since the bessctl command parser sets unpresent args to `None`

sangjinhan commented on core/worker.cc:
```diff
@@ -403,20 +406,59 @@ bool detach_tc(bess::TrafficClass *c) {
 
-WorkerPauser:: ~WorkerPauser() {
+WorkerPauser::~WorkerPauser() {
+  if (workers_paused_.size()) {
```
nit: `!workers_paused_.empty()`

sangjinhan commented on core/resume_hook.cc:
```diff
@@ -0,0 +1,70 @@
+std::set<Module *> event_modules;
+
+std::map<std::string, ResumeHookFactory>
+    &ResumeHookFactory::all_resume_hook_factories_holder(bool reset) {
```
We don't need `reset` here..? (I think this idion was used for gtest somewhere else, but not quite necessary here)

melvinw replied:
Yeah, it's carryover from ModuleBuilder to facilitate unit testing-- will remove.

sangjinhan commented on core/module.h:
```diff
@@ -287,6 +288,14 @@ class alignas(64) Module {
+   * whenever `e` occours. See `event.h` for details about the various event
+   * types and their semantics/requirements when it comes to modules.
+   */
+  virtual int OnEvent(bess::Event) { return ENOTSUP; }
```
To be consistent with other code, please consider using `-ENOTSUP`. The remaining C-style code expects int functions return negative numbers to report errors.

melvinw replied:
Sure thing

sangjinhan commented on core/resume_hooks/README.md:
```diff
@@ -0,0 +1,12 @@
+The API is simple, similar to that of `GateHook`. All you need to do is:
+
+- Define `YourHook::kName`. This must be unique across all other resume hooks.
+- Define `YourHook::kPriority`. Ties are broken by hook name in increasing lexographical order.
```
It may be worth mentioning that lower numbers have higher priority

melvinw commented:
Comments addressed. Not sure if [this weirdness](https://github.com/NetSys/bess/pull/661/files#diff-9e373311c9fa90b6c7c3412dfc306367R48) is okay, though

nrdmao commented on core/resume_hook.h:
```diff
@@ -0,0 +1,137 @@
+  virtual void Run() = 0;
+
+  bool operator<(const ResumeHook &rhs) const {
+    return priority_ < rhs.priority_ && name_ < rhs.name_;
```
Don't you want something like?:
return priority_ < rhs.priority_ || (priority_ == rhs.priority_ && name_ < rhs.name_);

nrdmao replied:
same issue in gate.cc

chris3torek replied:
I thought @melvinw had used `std::tie` to do this comparison, but maybe that was a different commit. But yes, when you can `std::tie` the various fields and compare, that's the way to go.

melvinw replied:
Ah, yup. After reading the definition of `operator<` on `std::tuple`, `std::tie` is definitely what we should use here. I switched to the current definition because I thought tuple comparisons were just `lhs[0] < rhs[0] && ... && lhs[n] < rhs[n]` and this was more explicit. Turns out tuples are actually ordered by `(lhs[0] < rhs[0] || !(rhs[0] < lhs[0])) && ...`

melvinw replied:
Fixed

nrdmao commented on core/worker.cc:
```diff
@@ -407,16 +410,31 @@ WorkerPauser::WorkerPauser() {
+  if (!workers_paused_.empty()) {
+    bess::run_global_resume_hooks(false);
+  }
+
```
I noticed in my testing that there should actually be a call made to attach_orphans() made before workers are resumed as is done in when the ResumeAll command is processed. Otherwise, if there are no subsequent calls made to ResumeAll when no workers are running, orphans will remain unattached. So, you can either add the call to attach_orphans() here or I can provide a fix after this code merges.

melvinw replied:
Great catch. I can fold the change into this PR

melvinw replied:
Fixed

sangjinhan commented:
(rebased)

sangjinhan commented:
Merged, thanks! Sorry for the delay.

![a-dar-tuesday-32](https://user-images.githubusercontent.com/1916976/31795129-807fd94a-b4d9-11e7-9aaf-ad54c7dfe290.jpg)
